qiskit/circuit/commutation_checker.py

```diff
@@ -37,32 +41,12 @@ class CommutationChecker:
     evicting from the cache less useful entries, etc.
     """
 
-    def __init__(self):
+    def __init__(self, cache_max_entries: int = 10**6):
         super().__init__()
-        self.cache = {}
-
-    def _hashable_parameters(self, params):
-        """Convert the parameters of a gate into a hashable format for lookup in a dictionary.
-
-        This aims to be fast in common cases, and is not intended to work outside of the lifetime of a
-        single commutation pass; it does not handle mutable state correctly if the state is actually
-        changed."""
-        try:
-            hash(params)
-            return params
-        except TypeError:
-            pass
-        if isinstance(params, (list, tuple)):
-            return tuple(self._hashable_parameters(x) for x in params)
-        if isinstance(params, np.ndarray):
-            # We trust that the arrays will not be mutated during the commutation pass, since nothing
-            # would work if they were anyway. Using the id can potentially cause some additional cache
-            # misses if two UnitaryGate instances are being compared that have been separately
-            # constructed to have the same underlying matrix, but in practice the cost of string-ifying
-            # the matrix to get a cache key is far more expensive than just doing a small matmul.
-            return (np.ndarray, id(params))
-        # Catch anything else with a slow conversion.
-        return ("fallback", str(params))
+        self._standard_commutations = StandardGateCommutations
+        self._cached_commutations = {}
+        self._cache_max_entries = cache_max_entries
```

Comment:
I like the idea to have the limit on the maximum number of cached entries. If I understand correctly, currently if the limit is reached the full cache would be cleared. I guess in theory we could add tracking for how many times a specific entry was checked and evict less useful entries, but this can be done in a follow-up PR (and only if you think it's worthwhile to do this).

Reply:
Yes, that is behaviour. I think what we should do ideally is to use the `lru_cache` decorator for the `_commute_matmul` function. However, I see two obstacles:
- We would need to get the gate object from the gate name as the `Operation` object is not hashable for some gates.
- We would need to pass the hashed version of the gate parameters in `_commute_matmul` and then reconstruct the gate parameter values inside that function. This is because the lru_cache decorator does not allow to set a custom key for a function signature  or to omit some of the parameters. At least not in Python 3.7. It should be possible with `cachetools` but I don't think it makes sense to add a new requirement to qiskit because of this. :-) I'm also not in favour of developing an LRU cache ourselves...